src/ess.jl

```diff
@@ -0,0 +1,353 @@
+
+    nothing
+end
+
+# estimation of the autocorrelation function
+function mean_autocorr(k::Int, cache::ESSCache)
+    # check arguments
+    samples = cache.samples
+    niter, nchains = size(samples)
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    var = cache.var
+    s = zero(eltype(var))
+    firstrange = 1:(niter - k)
+    lastrange = (k + 1):niter
+    @inbounds for i in 1:nchains
+        # increment unnormalized correlation estimates
+        if eltype(samples) isa LinearAlgebra.BlasReal
+            # call into BLAS if possible
+            s += dot(samples, firstrange, samples, lastrange) / var[i]
+            firstrange = firstrange .+ niter
+            lastrange = lastrange .+ niter
+        else
+            # otherwise use views
+            s += dot(view(samples, firstrange, i), view(samples, lastrange, i)) / var[i]
+        end
+    end
+
+    return s / length(samples)
+end
+
+function mean_autocorr(k::Int, cache::FFTESSCache)
+    # check arguments
+    niter = cache.niter
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    A = cache.A
+    nchains = size(A, 2)
+    s = zero(real(eltype(A)))
+    @inbounds for i in 1:nchains
+        s += real(A[k + 1, i]) / real(A[1, i])
+    end
+
+    return s / nchains
+end
+
+function mean_autocorr(k::Int, cache::BDAESSCache)
+    # check arguments
+    samples = cache.samples
+    niter, nchains = size(samples)
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    var = cache.var
+    s = zero(eltype(var))
+    @inbounds for j in 1:nchains
+        sj = zero(s)
+        for i in 1:(niter - k)
+            sj += abs2(samples[i, j] - samples[k + i, j])
+        end
+        s += sj / var[j]
+    end
+
+    return 1 - s / (2 * length(samples))
+end
+
+"""
+    ess(chains::Chains; sections = _default_sections(chains), kwargs...)
+
+Estimate the effective sample size and the potential scale reduction.
+"""
+function ess(
+    chains::Chains;
+    sections = _default_sections(chains), # perhaps directly nothing by default - or no sections argument?
+    kwargs...
+)
+    # subset the chain
+    _chains = Chains(chains, _clean_sections(chains, sections)) # may be unnecessary, already applied upstream
+
+    # estimate the effective sample size and rhat
+    ess, rhat = ess_rhat(_chains.value.data; kwargs...)
+
+    # convert to namedtuple
+    nt = merge((parameters = names(_chains),), (ess = ess, rhat = rhat))
+
+    return ChainDataFrame("ESS", nt)
+end
+
+function ess_rhat(chains::AbstractArray{<:Union{Missing,Real},3};
+                  method::AbstractESSMethod = ESSMethod(), kwargs...)
+    return ess_rhat(method, chains; kwargs...)
+end
+
+function ess_rhat(method, chains::AbstractArray{<:Union{Missing,Real},3};
+                  maxlag::Int = 250)
```

Comment:
Just noticed that we could just define the keyword argument version
```suggestion
function ess_rhat(
    chains::AbstractArray{<:Union{Missing,Real},3};
    method::AbstractESSMethod = ESSMethod(),
    maxlag::Int = 250
)
```
since we don't dispatch on `method`. Julia will still compile optimized code for each method.